### PR TITLE
Oracle DB Tests

### PR DESCRIPTION
**Pull request by ryanemerson**
https://issues.redhat.com/browse/ISPN-16404
https://issues.redhat.com/browse/ISPN-16587
https://issues.redhat.com/browse/ISPN-16588
https://issues.redhat.com/browse/ISPN-16591

### CODE REVIEW

**Comment by tristantarrant** [comment]
Can't you break the Oracle bits out into a separate PR ?

**Comment by ryanemerson** [comment]
@tristantarrant It seems the issue with adding Oracle tests is two fold:

1. The Jenkins agent doesn't have enough resources
2. The Oracle tests take longer due to the DB startup time causing the job to timeout.

I added a commit to restrict the DB matrix, as well as increasing the Jenkins time (via replay) and we now have a green build. The current approach for testing DBs is that we start all of the DB containers before executing the tests, but clearly this doesn't scale as we need to increase the resources available to the Jenkins agents and we're unable to do this with GH runners.

An alternative approach, which would require minimal code changes and speed up CI time, would be to execute DB specific tests in GH actions using a job matrix. The idea would be to exclude the DB related tests from the "normal" CI run of server/tests, then utilise a [GH matrix](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/running-variations-of-jobs-in-a-workflow) to execute a job performing the relevant tests for each DB.

**Comment by tristantarrant** [comment]
I'd rather fix our tests to start a single database at a time

**Comment by ryanemerson** [comment]
GH actions for mysql and oracle tests now in place and :green_circle:

An added bonus of this PR is that each PR build now uploads the server.zip as an artifact, which could be useful for further testing/debugging.

@tristantarrant  There were a few prerequisite steps required in order for `server/tests` to behave as required by the actions and I have tried to break these down into separate commits where possible so there purpose is clear.

@rigazilla Can you take a look at the GH actions?

File: .github/actions/build-infinispan/action.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+runs:
+  using: composite
+  steps:
+    - name: Setup Java
+      uses: actions/setup-java@v4
+      with:
+        java-version: 21
+        distribution: temurin
+        cache: maven
+
```
Comment (rigazilla) [inline .github/actions/build-infinispan/action.yml:13]: reuse the action defined below?

File: .github/workflows/on_propen_push_do_test.yml [hunk 2]
```diff
@@ -6,8 +6,24 @@ on:
...
       - 14.0.x
+
+concurrency:
+  # Only cancel jobs for PR updates
+  group: ci-${{ github.ref }}
+  cancel-in-progress: true
+
 jobs:
+
+  build:
```
Comment (rigazilla) [inline .github/workflows/on_propen_push_do_test.yml:17]: I will need to move this in a separate workflow if we want to build the images before the tests. but this is for a future step

**Review by rigazilla — CHANGES_REQUESTED** [review]

File: .github/workflows/on_propen_push_do_test.yml (hunk 2 above)
Comment (ryanemerson) [reply .github/workflows/on_propen_push_do_test.yml:17]: Sure. I kept the build part with the tests for simplicity, but we can easily split into multiple jobs.

**Review by rigazilla — APPROVED** [review]
github actions looks good to me!

File: .github/actions/build-infinispan/action.yml [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    - name: Setup Java
+      uses: ./.github/actions/setup-java
+
+    - name: Build Infinispan
+      shell: bash
+      run: |
+        # Ensure this plugin is built first to avoid warnings in the build
+        ./mvnw install -Pdistribution -s maven-settings.xml
+        # By using "dependency:resolve", it will download all dependencies used in later stages for running the tests
+        ./mvnw install dependency:resolve -V -e -DskipTests -DexcludeGroupIds=org.infinispan -Dsilent=true -s maven-settings.xml
```
Comment (pruivo) [inline .github/actions/build-infinispan/action.yml:16]: I don't understand this step. We are archiving only `.m2/repository/org/infinispan` but `dependency:resolve` will download other dependencies.

File: .github/actions/build-infinispan/action.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    - name: Compress Infinispan Maven artifacts
+      shell: bash
+      run: |
+        tar -C ~/ --use-compress-program zstd -cf m2-infinispan.tzts \
+        --exclude '*.tar.gz' \
```
Comment (pruivo) [inline .github/actions/build-infinispan/action.yml:22]: shouldn't `zip` be excluded too?

File: .github/actions/build-infinispan/action.yml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        .m2/repository/org/infinispan
+
+    - name: Upload Infinispan Maven artifacts
+      uses: actions/upload-artifact@v4
+      with:
+        name: m2-infinispan.tzts
```
Comment (pruivo) [inline .github/actions/build-infinispan/action.yml:28]: is the `name` private to the workflow or is it possible the concurrent PRs running this action will overwrite the upload?

File: .github/workflows/on_propen_push_do_test.yml [hunk 6]
```diff
@@ -18,23 +34,18 @@ jobs:
...
           github-token: ${{ secrets.GITHUB_TOKEN }}
 
-      - run: ./mvnw --version
+      - name: Extract Maven Artifacts
+        uses: ./.github/actions/extract-maven-artifacts
 
-      - name: Maven build
-        run: >
-            ./mvnw clean install -s maven-settings.xml -B -e -DskipTests -Pnative
+      - run: ./mvnw --version
```
Comment (pruivo) [inline .github/workflows/on_propen_push_do_test.yml:48]: is this for debugging? To print the versions, you could add `-V` to the `mvn verify`.

File: server/test.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+reset
+set -e
+RELEASES=~/Releases
+ROOT=~/workspace/RedHat/infinispan/infinispan
```
Comment (pruivo) [inline server/test.sh:5]: where do these directories come from? Is this script used anywhere?

**Review by pruivo — COMMENTED** [review]
I have a couple of questions but it looks good to me.

File: server/test.sh (hunk 7 above)
Comment (ryanemerson) [reply server/test.sh:5]: Nope! That's a local file of mine that accidentally got included, removing.

File: .github/workflows/on_propen_push_do_test.yml (hunk 6 above)
Comment (ryanemerson) [reply .github/workflows/on_propen_push_do_test.yml:48]: I *think* so, this step already existed and I was trying to disturb the existing workflow as much as possible. I can remove and add `-V` to `mvn verify` though :+1:

**Comment by pruivo** [comment]
@ryanemerson I think `JGroupsJdbcPing2IT` is not stopping JGroups for some reason. Those are logged after the test finishes
```
15:04:29,422 ERROR [o.j.p.JDBC_PING2] fv-az975-694-12808: failed reading from the DB java.sql.SQLException: No suitable driver found for jdbc:sqlserver://172.17.0.3:1433;encrypt=true;trustServerCertificate=true
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:708)
	at java.sql/java.sql.DriverManager.getConnection(DriverManager.java:230)
	at org.infinispan.server.persistence.JGroupsJdbcPing2IT$DummyDataSource.getConnection(JGroupsJdbcPing2IT.java:186)
	at org.infinispan.server.persistence.JGroupsJdbcPing2IT$DummyDataSource.getConnection(JGroupsJdbcPing2IT.java:181)
	at org.jgroups.protocols.JDBC_PING2.getConnection(JDBC_PING2.java:308)
	at org.jgroups.protocols.JDBC_PING2.readFromDB(JDBC_PING2.java:219)
	at org.jgroups.protocols.JDBC_PING2.readAll(JDBC_PING2.java:203)
	at org.jgroups.protocols.FILE_PING.findMembers(FILE_PING.java:146)
	at org.jgroups.protocols.Discovery.callFindMembersInAllDiscoveryProtocols(Discovery.java:393)
	at org.jgroups.protocols.Discovery.findMembers(Discovery.java:240)
	at org.jgroups.protocols.Discovery.down(Discovery.java:434)
	at org.jgroups.protocols.FILE_PING.down(FILE_PING.java:141)
	at org.jgroups.protocols.MERGE3$InfoSender.run(MERGE3.java:424)
	at org.jgroups.util.TimeScheduler3$Task.run(TimeScheduler3.java:332)
	at org.jgroups.util.TimeScheduler3$RecurringTask.run(TimeScheduler3.java:366)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

File: .github/actions/build-infinispan/action.yml (hunk 5 above)
Comment (ryanemerson) [reply .github/actions/build-infinispan/action.yml:28]: I believe it's private to the workflow. If you look at the url of the uploaded artifacts, each one has a unique artifact id which is a sub-resource of the individual run. Keycloak uses the same name on each PR without issue.

File: .github/actions/build-infinispan/action.yml (hunk 3 above)
Comment (ryanemerson) [reply .github/actions/build-infinispan/action.yml:16]: The `setup-java` also caches maven dependencies, so  what happens in the subsequent jobs is that the `setup-java` cache is restored to `~/.m2` and then we explicitly overwrite the `org/infinispan` artifacts with `./.github/actions/extract-maven-artifacts`. 

This is a similar approach to Keycloak, however in their CI they don't use the built in cache of `setup-java`. Instead they manually create the m2 cache using the `actions/cache` and explicitly exclude `org/keycloak` dependencies from this. This is more verbose, but the advantage of this approach is that it means that a PR which still has a reference to a removed module will never pass.

The  more I think about it, the more I think we should be safe and follow the same approach as Keycloak.

File: .github/actions/build-infinispan/action.yml (hunk 4 above)
Comment (ryanemerson) [reply .github/actions/build-infinispan/action.yml:22]: I don't think we produce any `*.tar.gz`, so I have just replaced this with `*.zip`

**Comment by ryanemerson** [comment]
> I think `JGroupsJdbcPing2IT` is not stopping JGroups for some reason. Those are logged after the test finishes

I've pushed a new commit that should fix it.

**Review by pruivo — APPROVED** [review]

**Comment by karesti** [comment]
@ryanemerson I launched the build here again

**Comment by rigazilla** [comment]
merged, thanks @ryanemerson !

**Pull request merged**